packet/eip_controlplane_reconciliation.go

```diff
@@ -0,0 +1,167 @@
+*/
+type controlPlaneEndpointManager struct {
+	inProcess     bool
+	apiServerPort int
+	eipTag        string
+	instances     cloudInstances
+	deviceIPSrv   packngo.DeviceIPService
+	ipResSvr      packngo.ProjectIPService
+	projectID     string
+	httpClient    *http.Client
+	k8sclient     kubernetes.Interface
+}
+
+func (m *controlPlaneEndpointManager) name() string {
+	return "controlPlaneEndpointManager"
+}
+
+func (m *controlPlaneEndpointManager) init(k8sclient kubernetes.Interface) error {
+	m.k8sclient = k8sclient
+	klog.V(2).Info("controlPlaneEndpointManager.init(): enabling BGP on project")
+	return nil
+}
+
+func (m *controlPlaneEndpointManager) nodeReconciler() nodeReconciler {
+	return m.reconciler()
+}
+func (m *controlPlaneEndpointManager) serviceReconciler() serviceReconciler {
+	return nil
+}
+
+func (m *controlPlaneEndpointManager) reconciler() func(nodes []*v1.Node, remove bool) error {
+	return func(nodes []*v1.Node, remove bool) error {
+		klog.V(2).Info("controlPlaneEndpoint.reconcile: new reconciliation")
+		if m.inProcess {
+			klog.V(2).Info("controlPlaneEndpoint.reconcileNodes: already in process, not starting a new one")
+			return nil
+		}
+		m.inProcess = true
+		defer func() {
+			m.inProcess = false
+		}()
+		if m.eipTag == "" {
+			return errors.New("elastic ip tag is empty. Nothing to do")
+		}
+		ipList, _, err := m.ipResSvr.List(m.projectID)
+		if err != nil {
+			return err
+		}
+		controlPlaneEndpoint := ipReservationByTags([]string{m.eipTag}, ipList)
+		if controlPlaneEndpoint == nil {
+			// IP NOT FOUND nothing to do here.
+			klog.Error("elastic IP not found. Please verify you have one with the expected tag: %s", m.eipTag)
+			return err
+		}
+		klog.Infof("healthcheck elastic ip %s", controlPlaneEndpoint.Address)
+		req, err := http.NewRequest("GET", fmt.Sprintf("https://%s:%d/healthz", controlPlaneEndpoint.Address, m.apiServerPort), nil)
+		if err != nil {
+			return err
+		}
+		resp, err := m.httpClient.Do(req)
+		if err != nil || resp.StatusCode != http.StatusOK {
+			if err := m.reassign(context.Background(), nodes, controlPlaneEndpoint); err != nil {
+				klog.Errorf("error reassigning control plane endpoint to a different device. err \"%s\"", err)
+				return err
+			}
+		}
+		defer resp.Body.Close()
+		return nil
+	}
+}
+
+func (m *controlPlaneEndpointManager) reassign(ctx context.Context, nodes []*v1.Node, ip *packngo.IPAddressReservation) error {
+	klog.V(2).Info("controlPlaneEndpoint.reassign")
+	for _, node := range nodes {
+		addresses, err := m.instances.NodeAddresses(ctx, types.NodeName(node.Name))
+		if err != nil {
+			return err
+		}
+		// It does not matter if the IP is external or internal. We can reach both.
+		a := addresses[0]
+
+		klog.Infof("healthcheck node %s", a.Address)
+		req, err := http.NewRequest("GET", fmt.Sprintf("https://%s:%d/healthz", a.Address, m.apiServerPort), nil)
+		if err != nil {
+			klog.Errorf("healthcheck failed for node %s. err \"%s\"", node.Name, err)
```

Comment:
This error still is a bit odd. The healthcheck did _not_ fail, and doesn't unless the `httpClient.Do()` fails. This is a URL construction error.